staging/src/k8s.io/apimachinery/pkg/labels/selector.go

```diff
@@ -867,25 +867,34 @@ func validateLabelValue(k, v string) error {
 	return nil
 }
 
-// SelectorFromSet returns a Selector which will match exactly the given Set. A
-// nil and empty Sets are considered equivalent to Everything().
+// SelectorFromSet is deprecated and unsafe, because it returns
+// a selector that matches everything on parsing errors such as too long values.
+// Use ValidatedSelectorFromSet or SelectorFromValidatedSet instead.
 func SelectorFromSet(ls Set) Selector {
-	if ls == nil || len(ls) == 0 {
+	selector, err := ValidatedSelectorFromSet(ls)
+	if err != nil {
 		return internalSelector{}
 	}
+	return selector
+}
+
+// ValidatedSelectorFromSet returns a Selector which will match exactly the given Set. A
+// nil and empty Sets are considered equivalent to Everything().
+func ValidatedSelectorFromSet(ls Set) (Selector, error) {
+	if ls == nil || len(ls) == 0 {
+		return internalSelector{}, nil
+	}
 	requirements := make([]Requirement, 0, len(ls))
 	for label, value := range ls {
 		r, err := NewRequirement(label, selection.Equals, []string{value})
-		if err == nil {
-			requirements = append(requirements, *r)
-		} else {
-			//TODO: double check errors when input comes from serialization?
-			return internalSelector{}
```

Comment:
Since this is a failure / problem condition, I don't see a problem with at least making it fail better, by returning as much of the selector as it could make.

Reply:
is it possible someone is checking for error conditions by seeing if the returned selector is empty? not endorsing such an approach, just wondering if that behavior change would be worse

Reply:
I don't see how this can fail, depending on how it does panicing might be acceptable, or maybe better yet, change the function signature to force people to fix their code.

Reply:
> is it possible someone is checking for error conditions by seeing if the returned selector is empty? not endorsing such an approach, just wondering if that behavior change would be worse

Yup, I remember we wrote some code that validates that the returned Selectors `Empty() == (len(map) == 0)` as a workaround after we fell hard on our faces with the default.

Reply:
(I created this PR after I had to build an "error handling" via `Empty()` again: https://github.com/kubernetes/test-infra/pull/16973/files#diff-163c29ff08fe88254e800a2264a84fe5R201)

> I don't see how this can fail, depending on how it does panicing might be acceptable, or maybe better yet, change the function signature to force people to fix their code.

This is in super wide use. IMHO panicing is not an option unless we make it always panic (as it may work for a long time and ppl will not notice and couldn't notice due to lack of go doc).

> change the function signature to force people to fix their code.

Yeah, that"s the idea, but we need to fix the code in this repo first, this is why I marked it as deprecated without removing it, because the fixing will not always be trivial.

Reply:
OK I think that's reasonable.